<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Unlocking single cell spatial omics analyses with scdney • scdneySpatialBiocAsia2024</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scdneySpatialBiocAsia2024</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/scdneySpatial.html">Unlocking single cell spatial omics analyses with scdney</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SydneyBioX/scdneySpatial_BiocAsia2024"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Unlocking single cell spatial omics analyses with scdney</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scdneySpatial_BiocAsia2024/blob/devel/vignettes/scdneySpatial.Rmd" class="external-link"><code>vignettes/scdneySpatial.Rmd</code></a></small>
      <div class="d-none name"><code>scdneySpatial.Rmd</code></div>
    </div>

    
    
<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
<p><strong>Presenting Authors</strong></p>
<p>Farhan
Ameen<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>,
Alex
Qin<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>,
Shreya
Rao<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>,
Ellis
Patrick<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>.</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>1</mn></msup><annotation encoding="application/x-tex">^1</annotation></semantics></math>
Westmead Institute for Medical Research, University of Sydney,
Australia<br><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>2</mn></msup><annotation encoding="application/x-tex">^2</annotation></semantics></math>
Sydney Precision Data Science Centre, University of Sydney,
Australia<br><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>3</mn></msup><annotation encoding="application/x-tex">^3</annotation></semantics></math>
School of Mathematics and Statistics, University of Sydney,
Australia</p>
<p><br> Contact: ellis.patrick@sydney.edu.au</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Understanding the interplay between different types of cells and
their immediate environment is critical for understanding the mechanisms
of cells themselves and their function in the context of human diseases.
Recent advances in high dimensional in situ cytometry technologies have
fundamentally revolutionized our ability to observe these complex
cellular relationships providing an unprecedented characterisation of
cellular heterogeneity in a tissue environment.</p>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>In this workshop we will introduce some of the key analytical
concepts needed to analyse data from high dimensional spatial omics
technologies such as, PhenoCycler, IMC, Xenium and MERFISH. We will show
how functionality from our Bioconductor packages simpleSeg, FuseSOM,
scClassify, scHot, spicyR, listClust, statial, scFeatures and ClassifyR
can be used to address various biological hypotheses. By the end of this
workshop attendees will be able to implement and assess some of the key
steps of a spatial analysis pipeline including cell segmentation,
feature normalisation, cell type identification, microenvironment and
cell-state characterisation, spatial hypothesis testing and patient
classification. Understanding these key steps will provide attendees
with the core skills needed to interrogate the comprehensive spatial
information generated by these exciting new technologies.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<p>It is expected that students will have:</p>
<ul>
<li>basic knowledge of R syntax,</li>
<li>this workshop will not provide an in-depth description of
cell-resolution spatial omics technologies.</li>
</ul>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<p>Several single cell R packages will be used from the scdney package,
for more information visit: <a href="https://sydneybiox.github.io/scdney/" class="external-link uri">https://sydneybiox.github.io/scdney/</a>.</p>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Understand and visualise spatial omics datasets.</li>
<li>Identify key biological questions that can be addressed with these
technologies and spatial analysis.</li>
<li>Understand the key analytical steps involved in spatial omics
analysis, and perform these steps using R.</li>
<li>Evaluate the performance of data normalisation and cell
segmentation.</li>
<li>Understand and generate individual feature representations from
spatial omics data.</li>
<li>Develop appreciation on how to assess performance of classification
models.</li>
<li>Perform disease outcome prediction using the feature representation
and robust classification framework.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="context">Context<a class="anchor" aria-label="anchor" href="#context"></a>
</h2>
<p>In the following we will re-analyse some IMC data <a href="https://doi.org/10.1158/1078-0432.CCR-22-1332" class="external-link">(Ferguson et al,
2022)</a> profiling the spatial landscape of head and neck cutaneous
squamous cell carcinomas (HNcSCC), the second most common type of skin
cancer. The majority of HNcSCC can be treated with surgery and good
local control, but a subset of large tumours infiltrate subcutaneous
tissue and are considered high risk for local recurrence and metastases.
The key conclusion of this manuscript (amongst others) is that spatial
information about cells and the immune environment can be used to
predict primary tumour progression or metastases in patients. We will
use our spicy workflow to reach a similar conclusion.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"simpleSeg"</span>, <span class="st">"cytomapper"</span>, <span class="st">"scClassify"</span>, <span class="st">"scHOT"</span>, <span class="st">"FuseSOM"</span>,<span class="st">"glmnet"</span>, <span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>,<span class="st">"Statial"</span>, <span class="st">"scFeatures"</span>, <span class="st">"ClassifyR"</span>, <span class="st">"tidyverse"</span>, <span class="st">"scater"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"STexampleData"</span>, <span class="st">"SpatialDatasets"</span>, <span class="st">"tidySingleCellExperiment"</span>, <span class="st">"scuttle"</span>, <span class="st">"batchelor"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># packages from scdney</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/scdneySpatial_BiocAsia2024/">scdneySpatialBiocAsia2024</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets" class="external-link">SpatialDatasets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/" class="external-link">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/" class="external-link">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial" class="external-link">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Other required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glmnet.stanford.edu" class="external-link">glmnet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">8</span>  <span class="co"># Feel free to parallelise things if you have the cores to spare.</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"utils.R"</span>, package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="download-images">Download images<a class="anchor" aria-label="anchor" href="#download-images"></a>
</h3>
<p>Feel free to follow along for this section. Much of the segmentation
pipeline is computationally demanding, and most likely won’t be able to
run on the Bioconductor Galaxy servers.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reads images to cache in a directory specified by the pathToImages variable.</span></span>
<span><span class="va">pathToImages</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html" class="external-link">Ferguson_Images</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Creates a temporary directory.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">pathToImages</span>, exdir <span class="op">=</span> <span class="va">tmp</span><span class="op">)</span> <span class="co"># Unzips files to the temporary directory.</span></span>
<span></span>
<span><span class="co"># Store images in a CytoImageList on_disk as h5 files to save memory.</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages</a></span><span class="op">(</span></span>
<span>  <span class="va">tmp</span>, <span class="co"># Read files from the temporary directory.</span></span>
<span>  single_channel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  on_disk <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h5FilesPath <span class="op">=</span> <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html" class="external-link">getHDF5DumpDir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean channel names</span></span>
<span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html" class="external-link">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="co"># Read in channel names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cn</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="va">cn</span><span class="op">)</span> <span class="co"># Remove preceding letters</span></span>
<span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".ome"</span>, <span class="st">""</span>, <span class="va">cn</span><span class="op">)</span> <span class="co"># Remove the .ome</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cn</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html" class="external-link">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cn</span> <span class="co"># Reassign channel names</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean image names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nam</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nam</span> <span class="co"># Reassigning image names</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">[[</span><span class="st">"imageID"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">nam</span> <span class="co"># Reassigning image names</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simpleseg-segmentation">SimpleSeg: Segmentation<a class="anchor" aria-label="anchor" href="#simpleseg-segmentation"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/simpleSeg.html" class="external-link">SimpleSeg
Package</a></p>
<div class="section level3">
<h3 id="run-simpleseg">Run simpleSeg<a class="anchor" aria-label="anchor" href="#run-simpleseg"></a>
</h3>
<p>If your images are stored in a <code>list</code> or
<code>CytoImageList</code> they can be segmented with a simple call to
<code><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg()</a></code>. To summarise, <code><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg()</a></code> is an R
implementation of a simple segmentation technique which traces out the
nuclei using a specified channel using <code>nucleus</code> then dilates
around the traced nuclei by a specified amount using
<code>discSize</code>. The nucleus can be traced out using either one
specified channel, or by using the principal components of all channels
most correlated to the specified nuclear channel by setting
<code>pca = TRUE</code>.</p>
<p>In the particular example below, we have asked <code>simpleSeg</code>
to do the following:</p>
<p>By setting <code>nucleus = c("HH3")</code>, we’ve asked simpleSeg to
trace out the nuclei signal in the images using the HH3 channel. By
setting <code>pca = TRUE</code>, simpleSeg segments out the nuclei mask
using a principal component analysis of all channels and using the
principal components most aligned with the nuclei channel, in this case,
HH3. By setting <code>cellBody = "dilate"</code>, simpleSeg uses a
dilation strategy of segmentation, expanding out from the nucleus by a
specified <code>discSize</code>. By setting <code>discSize = 3</code>,
simpleSeg dilates out from the nucleus by 3 pixels. By setting
<code>sizeSelection = 20</code>, simpleSeg ensures that only cells with
a size greater than 20 pixels will be used. By setting
<code>transform = "sqrt"</code>, simpleSeg square root transforms each
of the channels prior to segmentation. By setting
<code>tissue = c("panCK", "CD45", "HH3")</code>, we specify a tissue
mask which simpleSeg uses, filtering out all background noise outside
the tissue mask. This is important as these are tumour cores, wand hence
circular, so we’d want to ignore background noise which happens outside
of the tumour core.</p>
<p>There are many other parameters that can be specified in simpleSeg
(<code>smooth</code>, <code>watershed</code>, <code>tolerance</code>,
and <code>ext</code>), and we encourage the user to select the best
parameters which suit their biological context.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg</a></span><span class="op">(</span><span class="va">images</span>,</span>
<span>                   nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   pca <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   cellBody <span class="op">=</span> <span class="st">"dilate"</span>,</span>
<span>                   discSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   sizeSelection <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                   transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>                   tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"panCK"</span>, <span class="st">"CD45"</span>, <span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cores <span class="op">=</span> <span class="va">nCores</span></span>
<span>                   <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualise-outlines">Visualise outlines<a class="anchor" aria-label="anchor" href="#visualise-outlines"></a>
</h3>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> makes
it easy to overlay the mask on top of the nucleus intensity marker to
see how well our segmentation process has performed. Here we can see
that the segmentation appears to be performing reasonably.</p>
<p>If you see over or under-segmentation of your images,
<code>discSize</code> is a key parameter in <code><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg()</a></code> for
optimising the size of the dilation disc after segmenting out the
nuclei.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsee-seg-nuclei-1.png" style="border: 0px"></p>
<p>If you wish to visualise multiple markers instead of just the HH3
marker and see how the segmentation mask performs, this can also be
done. Here, we can see that our segmentation mask has done a good job of
capturing the CD31 signal, but perhaps not such a good job of capturing
the FXIIIA signal, which often lies outside of our dilated nuclear mask.
This could suggest that we might need to increase the
<code>discSize</code> of our dilation.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HH3"</span>, <span class="st">"CD31"</span>, <span class="st">"FX111A"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                         CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>                         FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span> <span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsee-seg-markers-1.png" style="border: 0px"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Is there any information we’re not capturing with this
segmentation?</li>
<li>What parameters might you change to improve the segmentation?</li>
<li>What are some intrinsic limitations of the simpleSeg method?</li>
</ol>
</div>
<p>In order to characterise the phenotypes of each of the segmented
cells, <code><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects()</a></code> from <code>cytomapper</code> will
calculate the average intensity of each channel within each cell as well
as a few morphological features. By default, the
<code><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects()</a></code> function will return a
<code>SingleCellExperiment</code> object, where the channel intensities
are stored in the <code>counts</code> assay and the spatial location of
each cell is stored in <code>colData</code> in the <code>m.cx</code> and
<code>m.cy</code> columns.</p>
<p>However, you can also specify <code><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects()</a></code> to return
a <code>SpatialExperiment</code> object by specifying
<code>return_as = "spe"</code>. As a <code>SpatialExperiment</code>
object, the spatial location of each cell is stored in the
<code>spatialCoords</code> slot, as <code>m.cx</code> and
<code>m.cy</code>, which simplifies plotting. In this demonstration, we
will return a <code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects</a></span><span class="op">(</span><span class="va">masks</span>,</span>
<span>                                    <span class="va">images</span>,</span>
<span>                                    img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                                    return_as <span class="op">=</span> <span class="st">"spe"</span>,</span>
<span>                                    BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoordsNames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span></code></pre></div>
<p>Next, to associate features in our image with disease progression, it
is important to read in information which links image identifiers to
their progression status. We will do this here, making sure that our
<code>imageID</code> column matches.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata/clinicalData_TMA1_2021_AF.csv"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="va">clinical</span><span class="op">[</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simpleseg-normalisation">simpleSeg: Normalisation<a class="anchor" aria-label="anchor" href="#simpleseg-normalisation"></a>
</h2>
<p>We’ve made it easy to follow on from the normalisation section. Our
<code>SpatialDatasets</code> package has the segmented SCE file stored
in the function <code><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html" class="external-link">spe_Ferguson_2022()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Download data now</span></span>
<span><span class="va">cells</span> <span class="op">=</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html" class="external-link">spe_Ferguson_2022</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>, <span class="st">"x"</span><span class="op">]</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>, <span class="st">"y"</span><span class="op">]</span></span></code></pre></div>
<p>Time for this code chunk to run: 2.23 seconds</p>
<div class="section level3">
<h3 id="normalise-the-data">Normalise the data<a class="anchor" aria-label="anchor" href="#normalise-the-data"></a>
</h3>
<p>We should check to see if the marker intensities of each cell require
some form of transformation or normalisation. The reason we do this is
two-fold:<br>
1) The intensities of images are often highly skewed, preventing any
meaningful downstream analysis.<br>
2) The intensities across different images are often different, meaning
that what is considered “positive” can be different across images.</p>
<p>By transforming and normalising the data, we aim to reduce these two
effects. Here we extract the intensities from the <code>counts</code>
assay. Looking at CD3 which should be expressed in the majority of the T
cells, the intensities are clearly very skewed, and it is difficult to
see what is considered a CD3- cell, and what is a CD3+ cell.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot densities of CD3 for each image.</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/cd3-density-1.png" width="480"></p>
<div class="question">
<p><strong>Questions</strong></p>
<p>Can we see what is a CD3+ vs a CD3- cell here?</p>
</div>
</div>
<div class="section level3">
<h3 id="dimension-reduction-and-visualisation">Dimension reduction and visualisation<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-visualisation"></a>
</h3>
<p>As our data is stored in a <code>SpatialExperiment</code> we can also
use <code>scater</code> to perform and visualise our data in a lower
dimension to look for batch effects in our images. We can see that
before normalisation, our UMAP shows a clear batch effect between
images.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Usually we specify a subset of the original markers which are informative to separating out distinct cell types for the UMAP and clustering.</span></span>
<span><span class="va">ct_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"podoplanin"</span>, <span class="st">"CD13"</span>, <span class="st">"CD31"</span>,</span>
<span>                <span class="st">"panCK"</span>, <span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8a"</span>,</span>
<span>                <span class="st">"CD20"</span>, <span class="st">"CD68"</span>, <span class="st">"CD16"</span>, <span class="st">"CD14"</span>, <span class="st">"HLADR"</span>, <span class="st">"CD66a"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a subset of images to plot.</span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/umap-1.png" width="700">Time
for this code chunk to run: 102.4 seconds</p>
<p>We can transform and normalise our data using the
<code>normalizeCells</code> function. In the
<code><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html" class="external-link">normalizeCells()</a></code> function, we specify the following
parameters. <code>transformation</code> is an optional argument which
specifies the function to be applied to the data. We do not apply an
arcsinh transformation here, as we already apply a square root transform
in the <code><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg()</a></code> function.
<code>method = c("trim99", "mean", PC1")</code> is an optional argument
which specifies the normalisation method/s to be performed. Here, we: 1)
Trim the 99th percentile 2) Divide by the mean 3) Remove the 1st
principal component <code>assayIn = "counts"</code> is a required
argument which specifies what the assay you’ll be taking the intensity
data from is named. In our context, this is called
<code>counts</code>.</p>
<p>This modified data is then stored in the <code>norm</code> assay by
default. We can see that this normalised data appears more bimodal, not
perfectly, but likely to a sufficient degree for clustering, as we can
at least observe a clear CD3+ peak at 1.00, and a CD3- peak at around
0.3.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Leave out the nuclei markers from our normalisation process. </span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DNA1"</span>, <span class="st">"DNA2"</span>, <span class="st">"HH3"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Transform and normalise the marker expression of each cell type.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html" class="external-link">normalizeCells</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                        markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>                        transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"mean"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>                        assayIn <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                        cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot densities of CD3 for each image</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"norm"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/normalisation-1.png" width="768">Time
for this code chunk to run: 8.1 seconds</p>
<div class="question">
<p><strong>Questions</strong></p>
<p>Can we see what is a CD3+ vs a CD3- cell here?</p>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"normUMAP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/norm-umap-1.png" width="700">Time
for this code chunk to run: 106.65 seconds</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>What are we hoping to achieve with normalisation?</li>
<li>What has changed between the pre-normalisation and
post-normalisation correction UMAP?</li>
</ol>
</div>
</div>
</div>
<div class="section level2">
<h2 id="fusesom">FuseSOM<a class="anchor" aria-label="anchor" href="#fusesom"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/FuseSOM.html" class="external-link">FuseSOM
Package</a> <a href="https://doi.org/10.1093/bioadv/vbad141" class="external-link">FuseSOM
Article</a></p>
<p>To dissect more information about cell interactions with each other,
and cell-specific differences between groups, we next perform
clustering. The <code><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM()</a></code> function from the
<code>FuseSOM</code> package conveniently runs clustering on any
<code>SingleCellExperiment</code> object, by specifying the number of
clusters under <code>numClusters</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set seed.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate SOM and cluster cells into 10 groups</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Time for this code chunk to run: 3.29 seconds</p>
<p>We can also observe how reasonable our choice of <code>k = 10</code>
was, using the <code><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster()</a></code> and
<code><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot()</a></code> functions. Here we examine the Gap method, but
others such as Silhouette and Within Cluster Distance are also
available. We can see that there are elbow points in the gap statistic
at <code>k = 7</code>, <code>k = 10</code>, and <code>k = 11</code>.
We’ve specified <code>k = 10</code>, striking a good balance between the
number of clusters and the gap statistic.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster</a></span><span class="op">(</span><span class="va">cells</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot</a></span><span class="op">(</span><span class="va">cells</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/numcluster-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Was our choice of <code>k = 10</code> a good choice?</li>
<li>What other <code>k</code> might have made sense?</li>
<li>From a biological perspective, what would increasing <code>k</code>
do?</li>
</ol>
</div>
<div class="section level3">
<h3 id="attempt-to-interpret-the-phenotype-of-each-cluster">Attempt to interpret the phenotype of each cluster<a class="anchor" aria-label="anchor" href="#attempt-to-interpret-the-phenotype-of-each-cluster"></a>
</h3>
<p>We can begin the process of understanding what each of these cell
clusters are by using the <code>plotGroupedHeatmap</code> function from
<code>scater</code>. At the least, here we can see we capture all the
major immune populations that we expect to see, including the CD4 and
CD8 T cells, the CD20+ B cells, the CD68+ myeloid populations, the CD66+
granulocytes, the podoplanin+ epithelial cells, and the panCK+ tumour
cells.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise marker expression in each cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  features <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  block <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/cluster-heatmap-1.png" width="700"></p>
<p>Given domain-specific knowledge of the tumour-immune landscape, we
can go ahead and annotate these clusters as cell types given their
expression profiles.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>cellType <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_1"</span> <span class="op">~</span> <span class="st">"GC"</span>, <span class="co"># Granulocytes</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_2"</span> <span class="op">~</span> <span class="st">"MC"</span>, <span class="co"># Myeloid cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_3"</span> <span class="op">~</span> <span class="st">"SC"</span>, <span class="co"># Squamous cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_4"</span> <span class="op">~</span> <span class="st">"EP"</span>, <span class="co"># Epithelial cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_5"</span> <span class="op">~</span> <span class="st">"SC"</span>, <span class="co"># Squamous cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_6"</span> <span class="op">~</span> <span class="st">"TC_CD4"</span>, <span class="co"># CD4 T cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_7"</span> <span class="op">~</span> <span class="st">"BC"</span>, <span class="co"># B cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_8"</span> <span class="op">~</span> <span class="st">"EC"</span>, <span class="co"># Endothelial cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_9"</span> <span class="op">~</span> <span class="st">"TC_CD8"</span>, <span class="co"># CD8 T cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_10"</span> <span class="op">~</span> <span class="st">"DC"</span> <span class="co"># Dendritic cells</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We might also be interested in how these cell types are distributed
on the images themselves. Here we examine the distribution of clusters
on image F3, noting the healthy epithelial and endothelial structures
surrounded by tumour cells.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/visualise-imageF3-1.png" width="700"></p>
<p>We find it always useful to check the number of cells in each
cluster.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check cell type frequencies.</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">cellType</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     DC     BC     MC     GC TC_CD8 TC_CD4     EC     EP     SC </span></span>
<span><span class="co">#&gt;   2411   4322   5283   7993   8534  11753  14159  14170  87288</span></span></code></pre>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Does our distribution of cell types make sense?</li>
<li>What would we normally expect as the most expressed cell type and
the least expressed cell type for this dataset?</li>
</ol>
</div>
<p>We can also use the UMAP we computed earlier to visualise our data in
a lower dimension and see how well our annotated cell types cluster
out.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># UMAP by cell type</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellType"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/umap-celltype-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="spicyr-test-spatial-relationships">spicyR: Test spatial relationships<a class="anchor" aria-label="anchor" href="#spicyr-test-spatial-relationships"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/spicyR.html" class="external-link">spicyR
Package</a> <a href="https://doi.org/10.1093/bioinformatics/btac268" class="external-link">spicyR
paper</a></p>
<p>spicyR uses the L-function to determine localisation or dispersion
between cell types. The L-function measures attraction or dispersion
between cell types, with larger values (&gt; 0) suggesting increased
attraction, and lower values (&lt; 0) suggesting avoidance between cell
types. We specify a range of radii to calculate the L-function over and
set <code>sigma = 50</code> to account for tissue inhomogeneity.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html" class="external-link">spicy</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                  condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>                  cellTypeCol <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                  imageIDCol <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                  Rs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">*</span><span class="fl">10</span>,</span>
<span>                  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code></pre></div>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>What does an L-function value of 0 suggest?</li>
<li>How would you choose the optimal radii (<code>Rs</code>) for a given
dataset?</li>
</ol>
</div>
<p>To save time we have saved the output of the above code.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"spicyTest.rda"</span>, package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html" class="external-link">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;             intercept coefficient    p.value adj.pvalue   from     to</span></span>
<span><span class="co">#&gt; BC__EC     -1.1346644    8.304020 0.03391529  0.6931721     BC     EC</span></span>
<span><span class="co">#&gt; EC__BC     -1.5909700    7.919971 0.04134661  0.6931721     EC     BC</span></span>
<span><span class="co">#&gt; GC__TC_CD4 -7.6316673   11.005639 0.05047096  0.6931721     GC TC_CD4</span></span>
<span><span class="co">#&gt; TC_CD4__GC -7.5406782   11.077602 0.05709461  0.6931721 TC_CD4     GC</span></span>
<span><span class="co">#&gt; TC_CD4__EP -0.7117393    4.853317 0.07261120  0.6931721 TC_CD4     EP</span></span>
<span><span class="co">#&gt; EP__TC_CD4 -0.7070918    4.820661 0.07510118  0.6931721     EP TC_CD4</span></span>
<span><span class="co">#&gt; EP__TC_CD8 -3.1445464    5.779536 0.08047994  0.6931721     EP TC_CD8</span></span>
<span><span class="co">#&gt; TC_CD8__EP -2.9313474    5.688420 0.09522058  0.6931721 TC_CD8     EP</span></span>
<span><span class="co">#&gt; TC_CD4__EC  3.1984391    4.366122 0.09716499  0.6931721 TC_CD4     EC</span></span>
<span><span class="co">#&gt; EC__TC_CD4  3.1052061    4.263971 0.10461424  0.6931721     EC TC_CD4</span></span></code></pre>
<p>We can visualise these tests using the <code>signifPlot</code>
function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise which relationships are changing the most.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html" class="external-link">signifPlot</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/signifPlot-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong> 1. Which cell types appear to have
significant co-localistion/dispersion? 2. Does this align with what we
might expect to see in a biological context?</p>
</div>
<p><code>spicyR</code> also has functionality for plotting out
individual pairwise relationships.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html" class="external-link">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             from <span class="op">=</span> <span class="st">"BC"</span>, </span>
<span>             to <span class="op">=</span> <span class="st">"EC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/spicyBoxPlot-1.png" width="700"></p>
<p>Alternatively, we can look at the most differentially localised
relationship between progressors and non-progressors by specifying
<code>rank = 1</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html" class="external-link">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             rank <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/spicyBoxPlot%20with%20rank-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="lisaclust-find-cellular-neighbourhoods">lisaClust: Find cellular neighbourhoods<a class="anchor" aria-label="anchor" href="#lisaclust-find-cellular-neighbourhoods"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/lisaClust.html" class="external-link">lisaClust
Package</a> <a href="https://doi.org/10.1002/cyto.a.24729" class="external-link">lisaClust
paper</a></p>
<p><code>lisaClust</code> allows us to identify regions where spatial
associations between cell types is similar. Here we use the
<code>lisaClust</code> function to clusters cells into 5 regions with
distinct spatial ordering.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Time for this code chunk to run: 36.98 seconds</p>
<p>To save time the results of the above code have been saved and can be
loaded in.</p>
<div class="section level3">
<h3 id="region---cell-type-enrichment-heatmap">Region - cell type enrichment heatmap<a class="anchor" aria-label="anchor" href="#region---cell-type-enrichment-heatmap"></a>
</h3>
<p>We can interpret which spatial orderings the regions are quantifying
using the <code>regionMap</code> function. This plots the frequency of
each cell type in a region relative to what you would expect by
chance.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise the enrichment of each cell type in each region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">cells</span>, cellType <span class="op">=</span> <span class="st">"cellType"</span>, limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/regionMap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualise-regions">Visualise regions<a class="anchor" aria-label="anchor" href="#visualise-regions"></a>
</h3>
<p>By default, these identified regions are stored in the
<code>regions</code> column in the <code>colData</code> of our
<code>SpatialExperiment</code> object. We can quickly examine the
spatial arrangement of these regions using <code>ggplot</code> on image
F3.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/plotReducedDim-1.png" width="700"></p>
<p>While much slower, the <code>hatchingPlot</code> function can also be
used to view regions of co-localisation simultaneously with the cell
type calls.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  nbp <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/hatchingPlot-1.png" width="700">Time
for this code chunk to run: 67.9 seconds</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How could the choice of <code>k</code> affect our clustering
results?</li>
<li>Looking at the graph above, would a different value of
<code>k</code> be better?</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="test-for-association-with-progression">Test for association with progression<a class="anchor" aria-label="anchor" href="#test-for-association-with-progression"></a>
</h3>
<p>We can use the <code>colTest</code> function to test for associations
between the proportions of cells in each region and progression
status.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if the proportion of each region is associated</span></span>
<span><span class="co"># with progression status.</span></span>
<span><span class="va">testRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html" class="external-link">colTest</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">testRegion</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          mean in group NP mean in group P tval.t pval adjPval  cluster</span></span>
<span><span class="co">#&gt; region_3            0.590           0.640  -1.20 0.24    0.47 region_3</span></span>
<span><span class="co">#&gt; region_2            0.052           0.039   0.96 0.34    0.47 region_2</span></span>
<span><span class="co">#&gt; region_1            0.140           0.120   0.73 0.47    0.47 region_1</span></span>
<span><span class="co">#&gt; region_4            0.210           0.200   0.73 0.47    0.47 region_4</span></span></code></pre>
<p>Time for this code chunk to run: 0.21 seconds</p>
<div class="question">
<p><strong>Questions</strong></p>
<p>How could results from <code>lisaClust</code> be used in conjunction
with results from <code>spicyR</code>?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="statial">Statial<a class="anchor" aria-label="anchor" href="#statial"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/Statial.html" class="external-link">Statial
Package</a></p>
<div class="section level3">
<h3 id="spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation-">SpatioMark: Continuous changes in marker expression associated with
varying levels of localisation.<a class="anchor" aria-label="anchor" href="#spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation-"></a>
</h3>
<p>The first step in analysing these changes is to calculate the spatial
proximity (<code>getDistances</code>) of each cell to every cell type.
These values will then be stored in the <code>reducedDims</code> slot of
the <code>SingleCellExperiment</code> object under the names distances.
SpatioMark also provides functionality to look into proximal cell
abundance using the <code>getAbundance()</code> function, which is
further explored within the <code>Statial</code> package vignette</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getDistances.html" class="external-link">getDistances</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>  maxDist <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Time for this code chunk to run: 7.41 seconds</p>
<p>We can then visualise an example image, specified with
<code>image = "F3"</code> and a particular marker interaction with cell
type localisation. To visualise these changes, we specify two cell types
with the <code>from</code> and <code>to</code> parameters, and a marker
with the <code>marker</code> parameter (cell-cell-marker interactions).
Here, we specify the changes in the marker podoplanin in <code>SC</code>
as its localisation to <code>EP</code> increases or decreases, where we
can observe that podoplanin decreases as its distance to the dense clump
of tumour cells increases.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html" class="external-link">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"SC"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"EP"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"podoplanin"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $image</span></span></code></pre>
<p><img src="scdneySpatial_files/figure-html/plotStateChange-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $scatter</span></span></code></pre>
<p><img src="scdneySpatial_files/figure-html/plotStateChange-2.png" width="700"></p>
<p>SpatioMark aims to holistically uncover all such significant
relationships by looking at all interactions across all images. The
<code><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges()</a></code> function provided by Statial can be
expanded for this exact purpose - by not specifying cell types, a
marker, or an image, <code><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges()</a></code> will examine the
most significant correlations between distance and marker expression
across the entire dataset.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">state_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  assay <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">state_dist</span><span class="op">[</span><span class="va">state_dist</span><span class="op">$</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span>,<span class="op">]</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;       imageID primaryCellType otherCellType     marker          coef      tval</span></span>
<span><span class="co">#&gt; 51810      F3              SC            EP podoplanin -0.0006911103 -16.90036</span></span>
<span><span class="co">#&gt; 23703      F3              EP            GC      CXCR3  0.0012103305  15.04988</span></span>
<span><span class="co">#&gt; 23704      F3              EP            GC     pSTAT3  0.0012299816  14.79933</span></span>
<span><span class="co">#&gt; 51978      F3              SC            GC       PDL2  0.0008802428  13.12601</span></span>
<span><span class="co">#&gt; 23669      F3              EP            MC       CCR7  0.0011476888  13.32900</span></span>
<span><span class="co">#&gt; 23775      F3              EP        TC_CD8      CXCR3  0.0016741880  12.95156</span></span>
<span><span class="co">#&gt; 51981      F3              SC            GC       ICOS  0.0005186181  11.75830</span></span>
<span><span class="co">#&gt; 23696      F3              EP            GC      CADM1  0.0011398625  12.18028</span></span>
<span><span class="co">#&gt; 11426      F3              EC            GC       CD31  0.0010070252  12.40866</span></span>
<span><span class="co">#&gt; 23758      F3              EP            EC       TIM3  0.0033539928  12.05028</span></span>
<span><span class="co">#&gt;               pval          fdr</span></span>
<span><span class="co">#&gt; 51810 4.203881e-59 9.004712e-57</span></span>
<span><span class="co">#&gt; 23703 8.041089e-41 8.191646e-39</span></span>
<span><span class="co">#&gt; 23704 8.916789e-40 8.712976e-38</span></span>
<span><span class="co">#&gt; 51978 1.839014e-37 1.637728e-35</span></span>
<span><span class="co">#&gt; 23669 9.360248e-34 6.863845e-32</span></span>
<span><span class="co">#&gt; 23775 3.025912e-32 2.085497e-30</span></span>
<span><span class="co">#&gt; 51981 1.113579e-30 7.070268e-29</span></span>
<span><span class="co">#&gt; 23696 3.239532e-29 1.903862e-27</span></span>
<span><span class="co">#&gt; 11426 4.839241e-29 2.825645e-27</span></span>
<span><span class="co">#&gt; 23758 1.030900e-28 5.930114e-27</span></span></code></pre>
<p>Time for this code chunk to run: 6.16 seconds</p>
<p>The results from our SpatioMark outputs can be converted from a
<code>data.frame</code> to a <code>matrix</code>, using the
<code><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix()</a></code> function. Note, the choice of extracting
either the t-statistic or the coefficient of the linear regression can
be specified using the <code>column = "tval"</code> parameter, with the
coefficient being the default extracted parameter. We can see that with
SpatioMark, we get some features which are significant after adjusting
for FDR.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Preparing outcome vector</span></span>
<span><span class="va">outcome</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">state_dist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">distMat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.0001</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">distMat</span>, <span class="va">outcome</span>, type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;                   mean in group NP mean in group P tval.t    pval adjPval</span></span>
<span><span class="co">#&gt; EC__EC__CD13              -0.00240        -9.6e-04   -3.6 0.00086    0.27</span></span>
<span><span class="co">#&gt; SC__SC__VISTA             -0.00470        -1.7e-03   -3.3 0.00190    0.27</span></span>
<span><span class="co">#&gt; SC__TC_CD4__CXCR3         -0.00015         8.8e-04   -3.4 0.00230    0.27</span></span>
<span><span class="co">#&gt; SC__EP__Ki67              -0.00037        -7.3e-05   -3.2 0.00260    0.27</span></span>
<span><span class="co">#&gt; EC__SC__CD68               0.00029        -3.9e-04    3.2 0.00300    0.27</span></span>
<span><span class="co">#&gt; SC__BC__DNA1               0.02600        -1.4e-01    3.0 0.00480    0.27</span></span>
<span><span class="co">#&gt;                             cluster</span></span>
<span><span class="co">#&gt; EC__EC__CD13           EC__EC__CD13</span></span>
<span><span class="co">#&gt; SC__SC__VISTA         SC__SC__VISTA</span></span>
<span><span class="co">#&gt; SC__TC_CD4__CXCR3 SC__TC_CD4__CXCR3</span></span>
<span><span class="co">#&gt; SC__EP__Ki67           SC__EP__Ki67</span></span>
<span><span class="co">#&gt; EC__SC__CD68           EC__SC__CD68</span></span>
<span><span class="co">#&gt; SC__BC__DNA1           SC__BC__DNA1</span></span></code></pre>
<p>Time for this code chunk to run: 0.4 seconds</p>
</div>
<div class="section level3">
<h3 id="kontextual">Kontextual<a class="anchor" aria-label="anchor" href="#kontextual"></a>
</h3>
<p><a href="https://www.biorxiv.org/content/10.1101/2024.09.03.611109v1" class="external-link">Pre-print</a></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster cell type hierarchy</span></span>
<span><span class="va">hierarchy</span> <span class="op">&lt;-</span> <span class="fu">treekoR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/treekoR/man/getClusterTree.html" class="external-link">getClusterTree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                     clusters <span class="op">=</span> <span class="va">cells</span><span class="op">$</span><span class="va">cellType</span>,</span>
<span>                                     hierarchy_method <span class="op">=</span> <span class="st">"hopach"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">parentList</span> <span class="op">=</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getParentPhylo.html" class="external-link">getParentPhylo</a></span><span class="op">(</span><span class="va">hierarchy</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualise tree</span></span>
<span><span class="va">hierarchy</span><span class="op">$</span><span class="va">clust_tree</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/hierarchy-1.png" width="700"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add parent 4 and 5</span></span>
<span><span class="va">parentList</span><span class="op">$</span><span class="va">parent_4</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">parentList</span><span class="op">$</span><span class="va">parent_2</span>, <span class="va">parentList</span><span class="op">$</span><span class="va">parent_3</span><span class="op">)</span></span>
<span><span class="va">parentList</span><span class="op">$</span><span class="va">parent_5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">parentList</span><span class="op">$</span><span class="va">parent_1</span>, <span class="st">"EC"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">cellType</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create data frame with all pairwise combinations of cells</span></span>
<span><span class="va">treeDf</span> <span class="op">=</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/parentCombinations.html" class="external-link">parentCombinations</a></span><span class="op">(</span><span class="va">all</span>, parentList <span class="op">=</span> <span class="va">parentList</span><span class="op">)</span></span></code></pre></div>
<p>Time for this code chunk to run: 1.39 seconds</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kontext</span> <span class="op">=</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/Kontextual.html" class="external-link">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  parentDf <span class="op">=</span> <span class="va">treeDf</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>To save time we have saved the output of the results above.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kontext.rda"</span>, package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can use the <code><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html" class="external-link">spicyR::spicy</a></code> to test for associations
between the <code>Kontextual</code> values and progression status. To do
so we use the <code>alternateResult</code> argument.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Kontextual result into data matrix</span></span>
<span><span class="va">kontextMat</span> <span class="op">&lt;-</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">kontext</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace NAs with 0</span></span>
<span><span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">kontextMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># Test association with outcome</span></span>
<span><span class="va">kontextOutcome</span> <span class="op">=</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html" class="external-link">spicy</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  alternateResult <span class="op">=</span> <span class="va">kontextMat</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot results</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html" class="external-link">signifPlot</a></span><span class="op">(</span><span class="va">kontextOutcome</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/kontextOutcome-1.png" width="960">Time
for this code chunk to run: 8.29 seconds</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Compare this plot to the spicyR results previously using
<code>signifPlot(spicyTest)</code>, how do Kontextual results compare to
spicyR?</li>
<li>What are the interpretation differences between spicyR and
Kontextual?</li>
<li>Use the <code>plotKontextual()</code> function to visualise some
relationships (example below), is there anything that strikes out to
you?</li>
</ol>
</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html" class="external-link">signifPlot</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/viewingKontextual-1.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">relationship</span> <span class="op">=</span> <span class="st">"TC_CD4__EC__parent_5"</span></span>
<span><span class="va">image</span> <span class="op">=</span> <span class="st">"G4"</span></span>
<span></span>
<span><span class="va">kontext</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">test</span> <span class="op">==</span> <span class="va">relationship</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    imageID                 test     original  kontextual  r inhomL</span></span>
<span><span class="co">#&gt; 1       A2 TC_CD4__EC__parent_5   7.68160860   6.5066053 50  FALSE</span></span>
<span><span class="co">#&gt; 2       A3 TC_CD4__EC__parent_5   2.23350953   3.4434658 50  FALSE</span></span>
<span><span class="co">#&gt; 3       A4 TC_CD4__EC__parent_5  -0.08701334   6.3304807 50  FALSE</span></span>
<span><span class="co">#&gt; 4       A5 TC_CD4__EC__parent_5  -0.57639962   4.5311090 50  FALSE</span></span>
<span><span class="co">#&gt; 5       A6 TC_CD4__EC__parent_5  10.02403012   7.5150350 50  FALSE</span></span>
<span><span class="co">#&gt; 6       B1 TC_CD4__EC__parent_5  10.87585261  14.2134530 50  FALSE</span></span>
<span><span class="co">#&gt; 7       B2 TC_CD4__EC__parent_5   5.06002008   5.9931607 50  FALSE</span></span>
<span><span class="co">#&gt; 8      B21 TC_CD4__EC__parent_5  -0.39959694   0.1112766 50  FALSE</span></span>
<span><span class="co">#&gt; 9       B3 TC_CD4__EC__parent_5   5.23542623   8.0275823 50  FALSE</span></span>
<span><span class="co">#&gt; 10      B4 TC_CD4__EC__parent_5  -0.66071472  -0.1844264 50  FALSE</span></span>
<span><span class="co">#&gt; 11      B6 TC_CD4__EC__parent_5   9.16587133   8.8867521 50  FALSE</span></span>
<span><span class="co">#&gt; 12      C1 TC_CD4__EC__parent_5   9.40908266   8.2981051 50  FALSE</span></span>
<span><span class="co">#&gt; 13      C2 TC_CD4__EC__parent_5  11.04803399   2.7263300 50  FALSE</span></span>
<span><span class="co">#&gt; 14      C3 TC_CD4__EC__parent_5   9.22058958   9.5992764 50  FALSE</span></span>
<span><span class="co">#&gt; 15      C4 TC_CD4__EC__parent_5  13.43380132   5.4622749 50  FALSE</span></span>
<span><span class="co">#&gt; 16      D2 TC_CD4__EC__parent_5  10.12907090  12.3327148 50  FALSE</span></span>
<span><span class="co">#&gt; 17      D3 TC_CD4__EC__parent_5  42.43221968  19.7863298 50  FALSE</span></span>
<span><span class="co">#&gt; 18      D4 TC_CD4__EC__parent_5   6.50868799   5.7006945 50  FALSE</span></span>
<span><span class="co">#&gt; 19      D5 TC_CD4__EC__parent_5   9.78058152   8.1404195 50  FALSE</span></span>
<span><span class="co">#&gt; 20      D6 TC_CD4__EC__parent_5   4.28348294   7.1094799 50  FALSE</span></span>
<span><span class="co">#&gt; 21      E1 TC_CD4__EC__parent_5  14.76511580   2.7278164 50  FALSE</span></span>
<span><span class="co">#&gt; 22      F3 TC_CD4__EC__parent_5 -14.09044015 -17.7783051 50  FALSE</span></span>
<span><span class="co">#&gt; 23      F4 TC_CD4__EC__parent_5 -15.53759618 -18.9689276 50  FALSE</span></span>
<span><span class="co">#&gt; 24      F5 TC_CD4__EC__parent_5 -12.86245147 -14.6367869 50  FALSE</span></span>
<span><span class="co">#&gt; 25      G4 TC_CD4__EC__parent_5  -2.39530472   2.3617033 50  FALSE</span></span>
<span><span class="co">#&gt; 26      G5 TC_CD4__EC__parent_5   5.51623888  10.5583310 50  FALSE</span></span>
<span><span class="co">#&gt; 27      G6 TC_CD4__EC__parent_5   4.33164604   4.7053925 50  FALSE</span></span>
<span><span class="co">#&gt; 28      H1 TC_CD4__EC__parent_5   1.94802803   7.4361377 50  FALSE</span></span>
<span><span class="co">#&gt; 29      H2 TC_CD4__EC__parent_5  13.35930713  13.0588006 50  FALSE</span></span>
<span><span class="co">#&gt; 30      H3 TC_CD4__EC__parent_5   1.34443428   5.0182459 50  FALSE</span></span>
<span><span class="co">#&gt; 31      H4 TC_CD4__EC__parent_5   7.54306921   3.3522477 50  FALSE</span></span>
<span><span class="co">#&gt; 32      H5 TC_CD4__EC__parent_5  -5.21413323  -0.9181290 50  FALSE</span></span>
<span><span class="co">#&gt; 33      H6 TC_CD4__EC__parent_5   1.66543198   5.6357928 50  FALSE</span></span>
<span><span class="co">#&gt; 34      I1 TC_CD4__EC__parent_5   2.68026706   9.4393784 50  FALSE</span></span>
<span><span class="co">#&gt; 35      I2 TC_CD4__EC__parent_5   9.77724059  16.0998082 50  FALSE</span></span>
<span><span class="co">#&gt; 36      I3 TC_CD4__EC__parent_5  33.93318280   7.6225714 50  FALSE</span></span>
<span><span class="co">#&gt; 37      I4 TC_CD4__EC__parent_5  32.63243351  19.1496354 50  FALSE</span></span>
<span><span class="co">#&gt; 38      I5 TC_CD4__EC__parent_5  25.11060806   9.8891406 50  FALSE</span></span>
<span><span class="co">#&gt; 39      J3 TC_CD4__EC__parent_5  23.47187850  23.5485701 50  FALSE</span></span>
<span><span class="co">#&gt; 40      J4 TC_CD4__EC__parent_5  23.69363791  22.2011439 50  FALSE</span></span>
<span><span class="co">#&gt; 41      J5 TC_CD4__EC__parent_5   6.56434930   5.9825165 50  FALSE</span></span>
<span><span class="co">#&gt; 42      J6 TC_CD4__EC__parent_5   2.98025405   2.9704760 50  FALSE</span></span>
<span><span class="co">#&gt; 43      K1 TC_CD4__EC__parent_5  -1.60932447   0.2013516 50  FALSE</span></span>
<span><span class="co">#&gt; 44      K2 TC_CD4__EC__parent_5   4.00813298   4.2639758 50  FALSE</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function from inst/extdata/utils.R file</span></span>
<span><span class="fu">plotKontextual</span><span class="op">(</span><span class="va">relationship</span>, imageChoose <span class="op">=</span> <span class="va">image</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/viewingKontextual-2.png" width="700">Time
for this code chunk to run: 0.91 seconds</p>
</div>
</div>
<div class="section level2">
<h2 id="classifyr-patient-classification">ClassifyR: Patient classification<a class="anchor" aria-label="anchor" href="#classifyr-patient-classification"></a>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html" class="external-link">ClassifyR
Package</a></p>
<p><a href="https://academic.oup.com/bioinformatics/article/31/11/1851/2365648?login=false" class="external-link">Journal
article</a></p>
<p>Our ClassifyR package formalises a convenient framework for
evaluating classification in R. We provide functionality to easily
include four key modelling stages; Data transformation, feature
selection, classifier training and prediction; into a cross-validation
loop. Here we use the <code>crossValidate</code> function to perform 50
repeats of 5-fold cross-validation to evaluate the performance of a
model (<code>classifier</code>) with feature selection
(<code>selectionMethod</code>) applied to five quantifications of our
IMC data; 1) Cell type proportions 2) Cell type localisation from
<code>spicyR</code> 3) Region proportions from <code>lisaClust</code> 4)
Cell type localisation in reference to a parent cell type from
<code>Kontextual</code> 5) Cell changes in response to proximal changes
from <code>SpatioMark</code></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create list to store data.frames</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add proportions of each cell type in each image</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Proportions"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"cellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add pair-wise associations</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html" class="external-link">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span>
<span><span class="va">spicyMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">spicyMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpicyR"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span></span>
<span></span>
<span></span>
<span><span class="co"># Add proportions of each region in each image to the list of dataframes.</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"LisaClust"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Add SpatioMark features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpatioMark"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">distMat</span></span>
<span></span>
<span><span class="co"># Add Kontextual features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Kontextual"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">kontextMat</span></span></code></pre></div>
<p>Time for this code chunk to run: 0.11 seconds</p>
<p>We have saved the data list and outcomes for convenience.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data.rda"</span>, package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform 5-fold cross-validation with 50 repeats.</span></span>
<span><span class="va">cv</span> <span class="op">&lt;-</span> <span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>  selectionMethod <span class="op">=</span> <span class="st">"t-test"</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"LASSOGLM"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nFeatures <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We have saved the results for convenience.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"cv.rda"</span>, package <span class="op">=</span> <span class="st">"scdneySpatialBiocAsia2024"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualise-cross-validated-prediction-performance">Visualise cross-validated prediction performance<a class="anchor" aria-label="anchor" href="#visualise-cross-validated-prediction-performance"></a>
</h3>
<p>Here we use the <code>performancePlot</code> function to assess the
AUC from each repeat of the 5-fold cross-validation.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Proportions"</span>, <span class="st">"SpicyR"</span>, <span class="st">"LisaClust"</span>, <span class="st">"Kontextual"</span>, <span class="st">"SpatioMark"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/performancePlot-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Run the following chunk of code, what is this plot telling us?</li>
<li>Do different methods perform differently on different samples?</li>
<li>Knowing this, what might we do to improve our model?</li>
</ol>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html" class="external-link">samplesMetricMap</a></span><span class="op">(</span><span class="va">cv</span><span class="op">)</span></span></code></pre></div>
<p><img src="scdneySpatial_files/figure-html/sampleMetrics-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; TableGrob (2 x 1) "arrange": 2 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name                 grob</span></span>
<span><span class="co">#&gt; 1 1 (2-2,1-1) arrange       gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,1-1) arrange text[GRID.text.2344]</span></span></code></pre>
<p>The following code chunk demonstrates how we might use information
from multiple matrices to perform classification.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DO NOT RUN</span></span>
<span><span class="va">multiCV</span> <span class="op">&lt;-</span> <span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>  selectionMethod <span class="op">=</span> <span class="st">"t-test"</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"LASSOGLM"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nFeatures <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  multiViewMethod <span class="op">=</span> <span class="st">"merge"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">multiCV</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fplot.png" style="border: 0px"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] scdneySpatialBiocAsia2024_2.0.1 knitr_1.48                     </span></span>
<span><span class="co">#&gt;  [3] glmnet_4.1-8                    Matrix_1.7-1                   </span></span>
<span><span class="co">#&gt;  [5] scater_1.34.0                   scuttle_1.16.0                 </span></span>
<span><span class="co">#&gt;  [7] cytomapper_1.18.0               EBImage_4.48.0                 </span></span>
<span><span class="co">#&gt;  [9] lubridate_1.9.3                 forcats_1.0.0                  </span></span>
<span><span class="co">#&gt; [11] stringr_1.5.1                   purrr_1.0.2                    </span></span>
<span><span class="co">#&gt; [13] readr_2.1.5                     tibble_3.2.1                   </span></span>
<span><span class="co">#&gt; [15] tidyverse_2.0.0                 ggplot2_3.5.1                  </span></span>
<span><span class="co">#&gt; [17] ttservice_0.4.1                 tidyr_1.3.1                    </span></span>
<span><span class="co">#&gt; [19] dplyr_1.1.4                     tidySingleCellExperiment_1.16.0</span></span>
<span><span class="co">#&gt; [21] ClassifyR_3.10.0                survival_3.7-0                 </span></span>
<span><span class="co">#&gt; [23] BiocParallel_1.40.0             MultiAssayExperiment_1.32.0    </span></span>
<span><span class="co">#&gt; [25] generics_0.1.3                  Statial_1.8.0                  </span></span>
<span><span class="co">#&gt; [27] spicyR_1.18.0                   lisaClust_1.14.0               </span></span>
<span><span class="co">#&gt; [29] FuseSOM_1.8.0                   simpleSeg_1.8.0                </span></span>
<span><span class="co">#&gt; [31] SpatialDatasets_1.4.0           SpatialExperiment_1.16.0       </span></span>
<span><span class="co">#&gt; [33] SingleCellExperiment_1.28.0     SummarizedExperiment_1.36.0    </span></span>
<span><span class="co">#&gt; [35] GenomicRanges_1.58.0            GenomeInfoDb_1.42.0            </span></span>
<span><span class="co">#&gt; [37] IRanges_2.40.0                  S4Vectors_0.44.0               </span></span>
<span><span class="co">#&gt; [39] MatrixGenerics_1.18.0           matrixStats_1.4.1              </span></span>
<span><span class="co">#&gt; [41] ExperimentHub_2.14.0            AnnotationHub_3.14.0           </span></span>
<span><span class="co">#&gt; [43] BiocFileCache_2.14.0            dbplyr_2.5.0                   </span></span>
<span><span class="co">#&gt; [45] Biobase_2.66.0                  BiocGenerics_0.52.0            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] tiff_0.1-12                 FCPS_1.3.4                 </span></span>
<span><span class="co">#&gt;   [3] nnet_7.3-19                 goftest_1.2-3              </span></span>
<span><span class="co">#&gt;   [5] Biostrings_2.74.0           HDF5Array_1.34.0           </span></span>
<span><span class="co">#&gt;   [7] TH.data_1.1-2               vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;   [9] spatstat.random_3.3-2       digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                   shape_1.4.6.1              </span></span>
<span><span class="co">#&gt;  [13] proxy_0.4-27                ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [15] deldir_2.0-4                permute_0.9-7              </span></span>
<span><span class="co">#&gt;  [17] magick_2.8.5                MASS_7.3-61                </span></span>
<span><span class="co">#&gt;  [19] pkgdown_2.1.1               reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.15               foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [23] withr_3.0.2                 ggfun_0.1.7                </span></span>
<span><span class="co">#&gt;  [25] psych_2.4.6.26              xfun_0.48                  </span></span>
<span><span class="co">#&gt;  [27] ggpubr_0.6.0                ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [29] memoise_2.0.1               ggbeeswarm_0.7.2           </span></span>
<span><span class="co">#&gt;  [31] RProtoBufLib_2.18.0         diptest_0.77-1             </span></span>
<span><span class="co">#&gt;  [33] princurve_2.1.6             systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt;  [35] tidytree_0.4.6              zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [37] GlobalOptions_0.1.2         ragg_1.3.3                 </span></span>
<span><span class="co">#&gt;  [39] V8_6.0.0                    DEoptimR_1.1-3             </span></span>
<span><span class="co">#&gt;  [41] prabclus_2.3-4              Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [43] KEGGREST_1.46.0             promises_1.3.0             </span></span>
<span><span class="co">#&gt;  [45] httr_1.4.7                  rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;  [47] rhdf5filters_1.18.0         fpc_2.2-13                 </span></span>
<span><span class="co">#&gt;  [49] rhdf5_2.50.0                UCSC.utils_1.2.0           </span></span>
<span><span class="co">#&gt;  [51] concaveman_1.1.0            curl_5.2.3                 </span></span>
<span><span class="co">#&gt;  [53] zlibbioc_1.52.0             ScaledMatrix_1.14.0        </span></span>
<span><span class="co">#&gt;  [55] analogue_0.17-7             polyclip_1.10-7            </span></span>
<span><span class="co">#&gt;  [57] GenomeInfoDbData_1.2.13     SparseArray_1.6.0          </span></span>
<span><span class="co">#&gt;  [59] fftwtools_0.9-11            doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [61] xtable_1.8-4                desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [63] evaluate_1.0.1              S4Arrays_1.6.0             </span></span>
<span><span class="co">#&gt;  [65] hms_1.1.3                   irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [67] colorspace_2.1-1            filelock_1.0.3             </span></span>
<span><span class="co">#&gt;  [69] spatstat.data_3.1-2         flexmix_2.3-19             </span></span>
<span><span class="co">#&gt;  [71] magrittr_2.0.3              ggtree_3.14.0              </span></span>
<span><span class="co">#&gt;  [73] later_1.3.2                 viridis_0.6.5              </span></span>
<span><span class="co">#&gt;  [75] modeltools_0.2-23           lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [77] robustbase_0.99-4-1         spatstat.geom_3.3-3        </span></span>
<span><span class="co">#&gt;  [79] XML_3.99-0.17               cowplot_1.1.3              </span></span>
<span><span class="co">#&gt;  [81] RcppAnnoy_0.0.22            ggupset_0.4.0              </span></span>
<span><span class="co">#&gt;  [83] class_7.3-22                svgPanZoom_0.3.4           </span></span>
<span><span class="co">#&gt;  [85] pillar_1.9.0                nlme_3.1-166               </span></span>
<span><span class="co">#&gt;  [87] iterators_1.0.14            compiler_4.4.1             </span></span>
<span><span class="co">#&gt;  [89] beachmat_2.22.0             stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [91] tensor_1.5                  minqa_1.2.8                </span></span>
<span><span class="co">#&gt;  [93] plyr_1.8.9                  treekoR_1.14.0             </span></span>
<span><span class="co">#&gt;  [95] crayon_1.5.3                abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [97] gridGraphics_0.5-1          locfit_1.5-9.10            </span></span>
<span><span class="co">#&gt;  [99] sp_2.1-4                    bit_4.5.0                  </span></span>
<span><span class="co">#&gt; [101] terra_1.7-83                sandwich_3.1-1             </span></span>
<span><span class="co">#&gt; [103] multcomp_1.4-26             fastcluster_1.2.6          </span></span>
<span><span class="co">#&gt; [105] codetools_0.2-20            textshaping_0.4.0          </span></span>
<span><span class="co">#&gt; [107] BiocSingular_1.22.0         bslib_0.8.0                </span></span>
<span><span class="co">#&gt; [109] coop_0.6-3                  GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt; [111] plotly_4.10.4               mime_0.12                  </span></span>
<span><span class="co">#&gt; [113] splines_4.4.1               circlize_0.4.16            </span></span>
<span><span class="co">#&gt; [115] Rcpp_1.0.13                 profileModel_0.6.1         </span></span>
<span><span class="co">#&gt; [117] blob_1.2.4                  utf8_1.2.4                 </span></span>
<span><span class="co">#&gt; [119] clue_0.3-65                 BiocVersion_3.20.0         </span></span>
<span><span class="co">#&gt; [121] lme4_1.1-35.5               fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [123] nnls_1.6                    ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt; [125] ggplotify_0.1.2             scam_1.2-17                </span></span>
<span><span class="co">#&gt; [127] statmod_1.5.0               tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt; [129] svglite_2.1.3               tweenr_2.0.3               </span></span>
<span><span class="co">#&gt; [131] pkgconfig_2.0.3             pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt; [133] tools_4.4.1                 cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [135] RSQLite_2.3.7               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [137] DBI_1.2.3                   numDeriv_2016.8-1.1        </span></span>
<span><span class="co">#&gt; [139] fastmap_1.2.0               rmarkdown_2.28             </span></span>
<span><span class="co">#&gt; [141] scales_1.3.0                grid_4.4.1                 </span></span>
<span><span class="co">#&gt; [143] shinydashboard_0.7.2        broom_1.0.7                </span></span>
<span><span class="co">#&gt; [145] sass_0.4.9                  patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [147] brglm_0.7.2                 BiocManager_1.30.25        </span></span>
<span><span class="co">#&gt; [149] carData_3.0-5               farver_2.1.2               </span></span>
<span><span class="co">#&gt; [151] mgcv_1.9-1                  yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [153] ggthemes_5.1.0              cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [155] hopach_2.66.0               lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [157] uwot_0.2.2                  mvtnorm_1.3-1              </span></span>
<span><span class="co">#&gt; [159] kernlab_0.9-33              backports_1.5.0            </span></span>
<span><span class="co">#&gt; [161] cytolib_2.18.0              timechange_0.3.0           </span></span>
<span><span class="co">#&gt; [163] gtable_0.3.6                rjson_0.2.23               </span></span>
<span><span class="co">#&gt; [165] ape_5.8                     parallel_4.4.1             </span></span>
<span><span class="co">#&gt; [167] limma_3.62.1                edgeR_4.4.0                </span></span>
<span><span class="co">#&gt; [169] jsonlite_1.8.9              bitops_1.0-9               </span></span>
<span><span class="co">#&gt; [171] bit64_4.5.2                 Rtsne_0.17                 </span></span>
<span><span class="co">#&gt; [173] FlowSOM_2.14.0              yulab.utils_0.1.7          </span></span>
<span><span class="co">#&gt; [175] vegan_2.6-8                 spatstat.utils_3.1-0       </span></span>
<span><span class="co">#&gt; [177] BiocNeighbors_2.0.0         ranger_0.16.0              </span></span>
<span><span class="co">#&gt; [179] flowCore_2.18.0             bdsmatrix_1.3-7            </span></span>
<span><span class="co">#&gt; [181] jquerylib_0.1.4             highr_0.11                 </span></span>
<span><span class="co">#&gt; [183] spatstat.univar_3.0-1       lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [185] ConsensusClusterPlus_1.70.0 shiny_1.9.1                </span></span>
<span><span class="co">#&gt; [187] diffcyt_1.26.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [189] rappdirs_0.3.3              glue_1.8.0                 </span></span>
<span><span class="co">#&gt; [191] XVector_0.46.0              RCurl_1.98-1.16            </span></span>
<span><span class="co">#&gt; [193] treeio_1.30.0               mclust_6.1.1               </span></span>
<span><span class="co">#&gt; [195] mnormt_2.1.1                coxme_2.2-22               </span></span>
<span><span class="co">#&gt; [197] jpeg_0.1-10                 gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [199] boot_1.3-31                 igraph_2.1.1               </span></span>
<span><span class="co">#&gt; [201] R6_2.5.1                    ggiraph_0.8.10             </span></span>
<span><span class="co">#&gt; [203] labeling_0.4.3              ggh4x_0.2.8                </span></span>
<span><span class="co">#&gt; [205] cluster_2.1.6               Rhdf5lib_1.28.0            </span></span>
<span><span class="co">#&gt; [207] aplot_0.2.3                 nloptr_2.1.1               </span></span>
<span><span class="co">#&gt; [209] DelayedArray_0.32.0         tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [211] vipor_0.4.7                 ggforce_0.4.2              </span></span>
<span><span class="co">#&gt; [213] raster_3.6-30               car_3.1-3                  </span></span>
<span><span class="co">#&gt; [215] AnnotationDbi_1.68.0        rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt; [217] munsell_0.5.1               DataVisualizations_1.3.2   </span></span>
<span><span class="co">#&gt; [219] data.table_1.16.2           ComplexHeatmap_2.22.0      </span></span>
<span><span class="co">#&gt; [221] htmlwidgets_1.6.4           RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [223] rlang_1.1.4                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">#&gt; [225] spatstat.explore_3.3-3      colorRamps_2.3.4           </span></span>
<span><span class="co">#&gt; [227] uuid_1.2-1                  lmerTest_3.1-3             </span></span>
<span><span class="co">#&gt; [229] ggnewscale_0.5.0            fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [231] beeswarm_0.4.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Farhan Ameen, Alex Qin, Shreya Rao, Ellis Patrick.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
